proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java

```diff
@@ -117,9 +118,11 @@ public void handle(PacketWrapper packet) throws Exception
     @Override
     public void handle(KeepAlive alive) throws Exception
     {
-        if ( alive.getRandomId() == con.getServer().getSentPingId() )
+        ServerConnection.KeepAliveInfo keepAliveInfo = con.getServer().getSentKeepAlives().getFirst();
+        if ( keepAliveInfo != null && alive.getRandomId() == keepAliveInfo.getId() )
         {
-            int newPing = (int) ( System.currentTimeMillis() - con.getSentPingTime() );
+            con.getServer().getSentKeepAlives().removeFirst(); // remove already peeked keepAliveInfo from queue since we recieved correct one
+            int newPing = (int) ( System.currentTimeMillis() - keepAliveInfo.getMillis() );
             con.getTabListHandler().onPingChange( newPing );
             con.setPing( newPing );
         } else
```

Comment:
Hmm, I think clients sending keepalives with a wrong id / in wrong order should also get disconnected.